File: pkg/directory/directory_test.go
package directory_test

import (
	"bytes"
	"path/filepath"
	"testing"

	"github.com/levis9527/jfs/pkg/directory"
	"github.com/levis9527/jfs/pkg/store"
)

func TestUploadGetDelete(t *testing.T) {
	root := t.TempDir()
	st, err := store.Open(store.Config{
		DataDir:     filepath.Join(root, "store"),
		VolumeCount: 2,
		VolumeSize:  1 << 20,
	})
	if err != nil {
		t.Fatal(err)
	}
	defer st.Close()

	dir, err := directory.Open(filepath.Join(root, "meta"), st, 1)
	if err != nil {
		t.Fatal(err)
	}
	defer dir.Close()

	data := []byte{0xff, 0xd8, 0xff, 0xe0} // fake jpeg header
	up, err := dir.Upload("img", "a.jpg", "image/jpeg", data)
	if err != nil {
		t.Fatal(err)
	}
	if up.Key == 0 || up.Vid == 0 {
		t.Fatalf("bad upload: %+v", up)
	}

	fm, got, err := dir.GetData("img", "a.jpg")
	if err != nil {
		t.Fatal(err)
	}
	if fm.Mime != "image/jpeg" || !bytes.Equal(got, data) {
		t.Fatalf("mismatch meta=%+v data=%v", fm, got)
	}

	if _, err := dir.Upload("img", "a.jpg", "image/jpeg", data); err != directory.ErrExists {
		t.Fatalf("want exists, got %v", err)
	}

	if err := dir.Delete("img", "a.jpg"); err != nil {
		t.Fatal(err)
	}
	if _, _, err := dir.GetData("img", "a.jpg"); err != directory.ErrNotFound {
		t.Fatalf("want not found, got %v", err)
	}
}

func TestMetadataPersistence(t *testing.T) {
	root := t.TempDir()
	storeDir := filepath.Join(root, "store")
	metaDir := filepath.Join(root, "meta")

	st, err := store.Open(store.Config{DataDir: storeDir, VolumeCount: 1, VolumeSize: 1 << 20})
	if err != nil {
		t.Fatal(err)
	}
	dir, err := directory.Open(metaDir, st, 2)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := dir.Upload("b", "f.bin", "application/octet-stream", []byte("xyz")); err != nil {
		t.Fatal(err)
	}
	_ = dir.Close()
	_ = st.Close()

	st2, err := store.Open(store.Config{DataDir: storeDir, VolumeCount: 1, VolumeSize: 1 << 20})
	if err != nil {
		t.Fatal(err)
	}
	defer st2.Close()
	dir2, err := directory.Open(metaDir, st2, 2)
	if err != nil {
		t.Fatal(err)
	}
	defer dir2.Close()
	_, got, err := dir2.GetData("b", "f.bin")
	if err != nil {
		t.Fatal(err)
	}
	if string(got) != "xyz" {
		t.Fatalf("got %q", got)
	}
}

File: src/test/java/com/levis9527/jfs/directory/DirectoryTest.java
package com.levis9527.jfs.directory;

import com.levis9527.jfs.store.Store;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class DirectoryTest {
    @TempDir
    Path temp;

    @Test
    void uploadGetDelete() throws Exception {
        Path root = temp.resolve("root");
        try (Store st = new Store(root.resolve("store"), 2, 1 << 20);
             Directory dir = new Directory(root.resolve("meta"), st, 1)) {
            byte[] data = new byte[]{(byte) 0xff, (byte) 0xd8, (byte) 0xff, (byte) 0xe0};
            var up = dir.upload("img", "a.jpg", "image/jpeg", data);
            assertTrue(up.key != 0);
            assertTrue(up.vid != 0);

            var got = dir.getData("img", "a.jpg");
            assertEquals("image/jpeg", got.meta().mime);
            assertArrayEquals(data, got.data());

            assertThrows(DirectoryException.class,
                    () -> dir.upload("img", "a.jpg", "image/jpeg", data));

            dir.delete("img", "a.jpg");
            assertThrows(DirectoryException.class, () -> dir.getData("img", "a.jpg"));
        }
    }

    @Test
    void metadataPersistence() throws Exception {
        Path storeDir = temp.resolve("store");
        Path metaDir = temp.resolve("meta");
        try (Store st = new Store(storeDir, 1, 1 << 20);
             Directory dir = new Directory(metaDir, st, 2)) {
            dir.upload("b", "f.bin", "application/octet-stream", "xyz".getBytes());
        }
        try (Store st2 = new Store(storeDir, 1, 1 << 20);
             Directory dir2 = new Directory(metaDir, st2, 2)) {
            assertArrayEquals("xyz".getBytes(), dir2.getData("b", "f.bin").data());
        }
    }

    @Test
    void overviewAndMetadataUpdate() throws Exception {
        Path root = temp.resolve("ov");
        try (Store st = new Store(root.resolve("store"), 2, 1 << 20);
             Directory dir = new Directory(root.resolve("meta"), st, 3)) {
            dir.upload("img", "a.jpg", "image/jpeg", new byte[]{1, 2, 3});
            dir.upload("doc", "n.txt", "text/plain", "hi".getBytes());

            var ov = dir.overview();
            assertEquals(2, ov.fileCount);
            assertEquals(2, ov.bucketCount);
            assertEquals(5, ov.totalBytes);
            assertEquals(2, ov.volumes.size());
            assertTrue(ov.volumes.get(0).usedBytes > 0 || ov.volumes.get(1).usedBytes > 0);

            var updated = dir.updateMeta("doc", "n.txt", "note.txt", "text/markdown");
            assertEquals("note.txt", updated.filename);
            assertEquals("text/markdown", updated.mime);
            assertEquals("text/markdown", dir.getMeta("doc", "note.txt").mime);
            assertThrows(DirectoryException.class, () -> dir.getMeta("doc", "n.txt"));
        }
    }

    @Test
    void renamePersistsAcrossRestart() throws Exception {
        Path storeDir = temp.resolve("store2");
        Path metaDir = temp.resolve("meta2");
        try (Store st = new Store(storeDir, 1, 1 << 20);
             Directory dir = new Directory(metaDir, st, 4)) {
            dir.upload("b", "old.bin", "application/octet-stream", "xyz".getBytes());
            dir.updateMeta("b", "old.bin", "new.bin", null);
        }
        try (Store st2 = new Store(storeDir, 1, 1 << 20);
             Directory dir2 = new Directory(metaDir, st2, 4)) {
            assertArrayEquals("xyz".getBytes(), dir2.getData("b", "new.bin").data());
            assertEquals(1, dir2.list(null).size());
            assertThrows(DirectoryException.class, () -> dir2.getMeta("b", "old.bin"));
        }
    }
}
